In [2]:
%load_ext sql

%sql postgresql://appdev@data/appdev

'Connected: appdev@appdev'

In [43]:
%sql SELECT * FROM pg_indexes WHERE tablename = 'circuits';

3 rows affected.


[('f1db', 'circuits', 'idx_17102_url', None, 'CREATE UNIQUE INDEX idx_17102_url ON circuits USING btree (url)'),
 ('f1db', 'circuits', 'circuits_position_idx', None, 'CREATE INDEX circuits_position_idx ON circuits USING gist ("position")'),
 ('f1db', 'circuits', 'idx_17102_primary', None, 'CREATE UNIQUE INDEX idx_17102_primary ON circuits USING btree (circuitid)')]

#### What type of indices exists for the table and why they are of that type (and not some other type)

As seen above there are 3 different indices found in this specific table.
first we have 2 Btree indices and 1 gist index.

###### idx_17102_url 
Is a btree because its a string that is indexed, and since btree works with simple logical operators like ">,<, =" it doesn't make sense to go for another type than btree for this index

###### idx_17102_primary
Is a btree and it has the same reasoning behind it being a btree as `idx_17102_url`, only difference is that this index is for integers instead of strings. But again it's a simple comparison

###### circuits_position_idx
Is a GiST the reason for choosing a GiST index here is because PostgreSQL's GiST implementation has logical operators that support two-dimensional geometric comparisons. 

#### The amount of space each index takes up

 * `idx_17102_url` takes up 16KB
 * `idx_17102_primary` also takes up 16KB 
 * `circuits_position_idx` takes up 8192 bytes or 8,192KB
All 3 indices size sum up to a bit over 40KB, This is actually a very small amount. 


#### We are talent scouts looking to win over some of the best new drivers there are
But we don't want them too old! Write a query that finds the winner of all the races, but only if they are younger than 38 years. The query should give return the date, driver surname, driver age, track time in milliseconds, race name and circuit name for all races.


In [53]:
%%sql 
SELECT  races.date AS race_date,
        milliseconds,
        drivers.surname,
        races.name AS race_name,
        circuits.name AS circuit_name,
        date_part('year', age(drivers.dob)) AS driver_age FROM results 
        JOIN drivers ON date_part('year', age(drivers.dob)) < 38 
            AND results.position = 1 
            AND drivers.driverid = results.driverid
        JOIN races USING (raceid) 
        JOIN circuits ON circuits.circuitid = races.circuitid;

184 rows affected.


[(datetime.date(2008, 3, 16), 5690616, 'Hamilton', 'Australian Grand Prix', 'Albert Park Grand Prix Circuit', 33.0),
 (datetime.date(2008, 4, 6), 5466970, 'Massa', 'Bahrain Grand Prix', 'Bahrain International Circuit', 36.0),
 (datetime.date(2008, 5, 11), 5209451, 'Massa', 'Turkish Grand Prix', 'Istanbul Park', 36.0),
 (datetime.date(2008, 5, 25), 7242742, 'Hamilton', 'Monaco Grand Prix', 'Circuit de Monaco', 33.0),
 (datetime.date(2008, 6, 8), 5784227, 'Kubica', 'Canadian Grand Prix', 'Circuit Gilles Villeneuve', 33.0),
 (datetime.date(2008, 6, 22), 5510245, 'Massa', 'French Grand Prix', 'Circuit de Nevers Magny-Cours', 36.0),
 (datetime.date(2008, 7, 6), 5949440, 'Hamilton', 'British Grand Prix', 'Silverstone Circuit', 33.0),
 (datetime.date(2008, 7, 20), 5480874, 'Hamilton', 'German Grand Prix', 'Hockenheimring', 33.0),
 (datetime.date(2008, 8, 3), 5847067, 'Kovalainen', 'Hungarian Grand Prix', 'Hungaroring', 36.0),
 (datetime.date(2008, 8, 24), 5732339, 'Massa', 'European Grand Prix', 'Valencia Street Circuit', 36.0),
 (datetime.date(2008, 9, 7), 4979394, 'Massa', 'Belgian Grand Prix', 'Circuit de Spa-Francorchamps', 36.0),
 (datetime.date(2008, 9, 14), 5207494, 'Vettel', 'Italian Grand Prix', 'Autodromo Nazionale di Monza', 30.0),
 (datetime.date(2008, 9, 28), 7036304, 'Alonso', 'Singapore Grand Prix', 'Marina Bay Street Circuit', 36.0),
 (datetime.date(2008, 10, 12), 5421892, 'Alonso', 'Japanese Grand Prix', 'Fuji Speedway', 36.0),
 (datetime.date(2008, 10, 19), 5517403, 'Hamilton', 'Chinese Grand Prix', 'Shanghai International Circuit', 33.0),
 (datetime.date(2008, 11, 2), 5651435, 'Massa', 'Brazilian Grand Prix', 'Autódromo José Carlos Pace', 36.0),
 (datetime.date(2007, 4, 8), 5534930, 'Alonso', 'Malaysian Grand Prix', 'Sepang International Circuit', 36.0),
 (datetime.date(2007, 4, 15), 5607515, 'Massa', 'Bahrain Grand Prix', 'Bahrain International Circuit', 36.0),
 (datetime.date(2007, 5, 13), 5496230, 'Massa', 'Spanish Grand Prix', 'Circuit de Barcelona-Catalunya', 36.0),
 (datetime.date(2007, 5, 27), 6029329, 'Alonso', 'Monaco Grand Prix', 'Circuit de Monaco', 36.0),
 (datetime.date(2007, 6, 10), 6251292, 'Hamilton', 'Canadian Grand Prix', 'Circuit Gilles Villeneuve', 33.0),
 (datetime.date(2007, 6, 17), 5469965, 'Hamilton', 'United States Grand Prix', 'Indianapolis Motor Speedway', 33.0),
 (datetime.date(2007, 7, 22), 7586358, 'Alonso', 'European Grand Prix', 'Nürburgring', 36.0),
 (datetime.date(2007, 8, 5), 5752991, 'Hamilton', 'Hungarian Grand Prix', 'Hungaroring', 33.0),
 (datetime.date(2007, 8, 26), 5202161, 'Massa', 'Turkish Grand Prix', 'Istanbul Park', 36.0),
 (datetime.date(2007, 9, 9), 4717806, 'Alonso', 'Italian Grand Prix', 'Autodromo Nazionale di Monza', 36.0),
 (datetime.date(2007, 9, 30), 7234579, 'Hamilton', 'Japanese Grand Prix', 'Fuji Speedway', 33.0),
 (datetime.date(2006, 3, 12), 5386205, 'Alonso', 'Bahrain Grand Prix', 'Bahrain International Circuit', 36.0),
 (datetime.date(2006, 4, 2), 5667870, 'Alonso', 'Australian Grand Prix', 'Albert Park Grand Prix Circuit', 36.0),
 (datetime.date(2006, 5, 14), 5181759, 'Alonso', 'Spanish Grand Prix', 'Circuit de Barcelona-Catalunya', 36.0),
 (datetime.date(2006, 5, 28), 6223116, 'Alonso', 'Monaco Grand Prix', 'Circuit de Monaco', 36.0),
 (datetime.date(2006, 6, 11), 5151927, 'Alonso', 'British Grand Prix', 'Silverstone Circuit', 36.0),
 (datetime.date(2006, 6, 25), 5677308, 'Alonso', 'Canadian Grand Prix', 'Circuit Gilles Villeneuve', 36.0),
 (datetime.date(2006, 8, 27), 5331082, 'Massa', 'Turkish Grand Prix', 'Istanbul Park', 36.0),
 (datetime.date(2006, 10, 8), 5032413, 'Alonso', 'Japanese Grand Prix', 'Suzuka Circuit', 36.0),
 (datetime.date(2006, 10, 22), 5513751, 'Massa', 'Brazilian Grand Prix', 'Autódromo José Carlos Pace', 36.0),
 (datetime.date(2005, 3, 20), 5493736, 'Alonso', 'Malaysian Grand Prix', 'Sepang International Circuit', 36.0),
 (datetime.date(2005, 4, 3), 5358531, 'Alonso', 'Bahrain Grand Prix', 'Bahrain International Circuit', 36.0),

In [54]:
%%sql 
EXPLAIN ANALYZE
SELECT  races.date AS race_date,
        milliseconds,
        drivers.surname,
        races.name AS race_name,
        circuits.name AS circuit_name,
        date_part('year', age(drivers.dob)) AS driver_age FROM results 
        JOIN drivers ON date_part('year', age(drivers.dob)) < 38 
            AND results.position = 1 
            AND drivers.driverid = results.driverid
        JOIN races USING (raceid) 
        JOIN circuits ON circuits.circuitid = races.circuitid;

22 rows affected.


[('Hash Join  (cost=82.13..815.57 rows=324 width=66) (actual time=1.066..4.306 rows=184 loops=1)',),
 ('  Hash Cond: (races.circuitid = circuits.circuitid)',),
 ('  ->  Hash Join  (cost=78.48..803.83 rows=324 width=50) (actual time=0.895..4.022 rows=184 loops=1)',),
 ('        Hash Cond: (results.raceid = races.raceid)',),
 ('        ->  Hash Join  (cost=38.53..759.78 rows=324 width=27) (actual time=0.564..3.650 rows=184 loops=1)',),
 ('              Hash Cond: (results.driverid = drivers.driverid)',),
 ('              ->  Seq Scan on results  (cost=0.00..708.96 rows=974 width=24) (actual time=0.004..2.886 rows=974 loops=1)',),
 ('                    Filter: ("position" = 1)',),
 ('                    Rows Removed by Filter: 22703',),
 ('              ->  Hash  (cost=35.03..35.03 rows=280 width=19) (actual time=0.552..0.552 rows=60 loops=1)',),
 ('                    Buckets: 1024  Batches: 1  Memory Usage: 12kB',),
 ('                    ->  Seq Scan on drivers  (cost=0.00..35.03 rows=280 width=19) (actual time=0.010..0.536 rows=60 loops=1)',),
 ("                          Filter: (date_part('year'::text, age((CURRENT_DATE)::timestamp with time zone, (dob)::timestamp with time zone)) < '38'::double precision)",),
 ('                          Rows Removed by Filter: 781',),
 ('        ->  Hash  (cost=27.76..27.76 rows=976 width=39) (actual time=0.322..0.322 rows=976 loops=1)',),
 ('              Buckets: 1024  Batches: 1  Memory Usage: 79kB',),
 ('              ->  Seq Scan on races  (cost=0.00..27.76 rows=976 width=39) (actual time=0.006..0.155 rows=976 loops=1)',),
 ('  ->  Hash  (cost=2.73..2.73 rows=73 width=28) (actual time=0.110..0.110 rows=73 loops=1)',),
 ('        Buckets: 1024  Batches: 1  Memory Usage: 13kB',),
 ('        ->  Seq Scan on circuits  (cost=0.00..2.73 rows=73 width=28) (actual time=0.076..0.089 rows=73 loops=1)',),
 ('Planning time: 0.620 ms',),
 ('Execution time: 4.394 ms',)]

#### Describe the query using ``EXPLAIN ANALYZE`` 

###### How many calls are you making?
There is being made 10 function calls before this query is done.
First it starts by joining the 3 different tables on the cinditions i have set.
 * Hash Cond: (races.circuitid = circuits.circuitid) 
 * Hash Cond: (results.raceid = races.raceid)
 * Hash Cond: (results.driverid = drivers.driverid)
When these 3 steps are done it runs a scan on the resultset. then it runs the Filter function on the result set and removes all results where position isn't equal to 1, we only want first place for each race.

after this it runs a scan on the drivers where we filter for age, we don't want anybody that is older than 38 years old. And after the filter function is done it removes the drivers who are older than 38 from the resultset and returns a new resultset.

Then lastly it runs a scan on both the `races` table and `circuits` to get the values that match our resultset.

###### How long does it take to perform the query?

The query had a Planning time of `0.620 ms`
and a Execution time of `4.394 ms`.



In [60]:
%%sql 
CREATE MATERIALIZED VIEW race_winners_underage_38 AS 
SELECT  races.date AS race_date,
        milliseconds,
        drivers.surname,
        races.name AS race_name,
        circuits.name AS circuit_name,
        date_part('year', age(drivers.dob)) AS driver_age FROM results 
        JOIN drivers ON date_part('year', age(drivers.dob)) < 38 
            AND results.position = 1 
            AND drivers.driverid = results.driverid
        JOIN races USING (raceid) 
        JOIN circuits ON circuits.circuitid = races.circuitid;

184 rows affected.


[]

In [62]:
%sql EXPLAIN ANALYZE SELECT * FROM race_winners_underage_38;

3 rows affected.


[('Seq Scan on race_winners_underage_38  (cost=0.00..10.50 rows=50 width=1568) (actual time=0.012..0.038 rows=184 loops=1)',),
 ('Planning time: 0.170 ms',),
 ('Execution time: 0.104 ms',)]

#### Create a materialized view of your query. Using ``EXPLAIN ANALYZE`` try to query the view.

The Planning time is `0.170 ms` Where before it was `0.620 ms`.
The Execution time is `0.104 ms` Where before it was `4.394 ms`.

From the look of the time it takes to get the result one would say that Views are better, but the problem with views is that it's basicly a snapshot of the resultset from when the view was created. This means if the database was updated the view would still return the same as it did before the update. So if you want the newest data at all times and want to make sure that you get it, it could make sense to run the query each time. But if it's data that rarely changes you can create a view.